jpype/_jproxy.py

```diff
@@ -22,46 +22,64 @@
 # FIXME the java.lang.method we are overriding should be passes to the lookup function
 # so we can properly handle name mangling on the override.
 
-
-def _createJProxy(cls, *intf, **kwargs):
-    """ (internal) Create a proxy from a python class with
-    @JOverride notation on methods.
-    """
-    # Convert the interfaces list
-    actualIntf = _convertInterfaces(intf)
-
-    # Find all class defined overrides
-    overrides = {}
-    for k, v in cls.__dict__.items():
-        try:
-            attr = object.__getattribute__(v, "__joverride__")
-            overrides[k] = (v, attr)
-        except AttributeError:
-            pass
-
+def _checkInterfaceOverrides(interfaces, overrides):
     # Verify all methods are overriden
-    for interface in actualIntf:
+    for interface in interfaces:
         for method in interface.class_.getMethods():
             if method.getModifiers() & 1024 == 0:
                 continue
             if not str(method.getName()) in overrides:
                 raise NotImplementedError("Interface '%s' requires method '%s' to be implemented." % (
                     interface.class_.getName(), method.getName()))
 
-    members = {}
+
+def _createJProxy(cls, *intf, **kwargs):
+    """ (internal) Create a proxy from a python class with
+    @JOverride notation on methods.
+    """
+    # TODO: When Python2 is dropped these kwargs can become keyword-only kwargs
+    deferred = bool(kwargs.pop('deferred', False))
+    if kwargs:
+        raise TypeError('Invalid keywords: {}'.format(
+            ','.join([str(kwarg) for kwarg in kwargs])))
+
+    def _classOverrides(cls):
+        # Find all class defined overrides
+        overrides = {}
+        for k, v in cls.__dict__.items():
+            try:
+                attr = object.__getattribute__(v, "__joverride__")
+                overrides[k] = (v, attr)
+            except AttributeError:
+                pass
+        return overrides
+
+    def _prepareInterfaces(cls, intf):
+        # Convert the interfaces list
+        actualIntf = _convertInterfaces(intf)
+        overrides = _classOverrides(cls)
+        _checkInterfaceOverrides(actualIntf, overrides)
+        return actualIntf
 
     def new(tp, *args, **kwargs):
-        self = _jpype._JProxy.__new__(tp, None, actualIntf)
-        cls.__init__(self, *args, **kwargs)
+        # Attach a __jpype_interfaces__ attribute to this class if
+        # one doesn't already exist.
+        if getattr(tp, "__jpype_interfaces__", None) is None:
```

Comment:
Having only one new method here rather than one for the deferred and the normal case means we are adding burden to the use of proxies.  My general philosophy is attemp not to slow an old feature to add a new one.  Not that they are very speedy currently, but one day they may need to be.  Also the point of a get with default is to assign to a var so that it can be a grab and go.  As currently written it would make two trips through the object lookup table.  I haven't done much benchmarking of byte codes on repeated look ups, but i haven't found a lot of optimization sections in my reading of cpython source so I have been assuming it is rather literal.  Perhaps i am wrong on this.